File: 0941-valid-mountain-array/0941-valid-mountain-array.cpp
class Solution {
public:
    bool validMountainArray(vector<int>& arr) {
        if(arr.size() == 1) return false;
        int cnt = 0;
        for(int i = 0; i < arr.size() - 1; i++){
            if(arr[i] < arr[i+1]){
                if(cnt == 0) cnt = 1;
                else if(cnt == -1) return false;
            }else if(arr[i] > arr[i+1]){
                if(cnt == 1) cnt = -1;
                else if(cnt == 0) return false;
            }else{
                return false;
            }
        }
        return cnt == -1;

        //  ==============================  //
        //  submission 제일 빠른 코드
        // int size = arr.size();
        // int i = 0;
        // while(i+1 < size && arr[i] < arr[i+1]){
        //     i++;
        // }
        // if(i == 0 || i == size - 1){
        //     return false;
        // }
        // while(i+1 < size && arr[i] > arr[i+1]){
        //     i++;
        // }
        // return i == size-1;
    }
};

File: 0941-valid-mountain-array/0941-valid-mountain-array.cs
public class Solution {
    public bool ValidMountainArray(int[] arr) {
        //  https://leetcode.com/problems/valid-mountain-array/solutions/3350244/c-solution
        int maxNum = arr.Max();
        //  가장 큰 수의 인덱스를 찾는다
        var maxIndex = Array.FindIndex(arr, n => n == maxNum);

        //  인덱스가 0이거나 제일 끝 인덱스라면 false를 반환
        //  증가하는 부분이 없거나 증가하는 부분만 있기 때문
        if(maxIndex == 0 || maxIndex == arr.Length-1){
            return false;
        }
        
        int currentRight = arr[maxIndex];
        //  가장 큰 값의 오른쪽이 내림차순이 맞는지 확인
        for(int i = maxIndex + 1; i < arr.Length; i++){
            bool ans = currentRight > arr[i];
            if(!ans) return false;
            currentRight = arr[i];
        }

        int currentLeft = arr[0];
        //  첫 번째 값부터 가장 큰 값까지 오름차순인지 확인
        for(int j = 1; j < maxIndex; j++){
            bool ans = currentLeft < arr[j];
            if(!ans) return false;
            currentLeft = arr[j];
        }

        return true;
    }
}